sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelOperationRuntimeGroupFiltering.scala

```diff
@@ -89,10 +88,8 @@ case class RowLevelOperationRuntimeGroupFiltering(optimizeSubqueries: Rule[Logic
       buildKeys: Seq[Attribute],
       pruningKeys: Seq[Attribute]): Expression = {
 
-    val buildQuery = Project(buildKeys, matchingRowsPlan)
-    val dynamicPruningSubqueries = pruningKeys.zipWithIndex.map { case (key, index) =>
-      DynamicPruningSubquery(key, buildQuery, buildKeys, index, onlyInBroadcast = false)
-    }
-    dynamicPruningSubqueries.reduce(And)
+    val buildQuery = Aggregate(buildKeys, buildKeys, matchingRowsPlan)
```

Comment:
Are there any downsides of rewriting `DynamicPruningSubquery` into `DynamicPruningExpression` directly instead of relying on other rules like `PlanDynamicPruningFilters` and `PlanAdaptiveDynamicPruningFilters`? I see some special branches for exchange reuse in those rules that would not apply now.

Reply:
Got it. I was originally worried we could miss some future optimizations given that dynamic pruning for row-level operations would go through a different route compared to the normal DPP.

One alternative could be to extend `DynamicPruningSubquery` with a flag whether it should be optimized or not. Up to you, though.

Reply:
My rationale is, what we really need is a subquery here. This is completely different from dynamic partition pruning. One limitation is DS v2 runtime filter pushdown only applies to `DynamicPruningExpression`. We can probably fix that and accept normal non-correlated subqueries as well.

Reply:
Yeah, DS v2 runtime filtering framework is fairly limited at this point.